src/sleep/sleep.c

```diff
@@ -263,40 +263,90 @@ static int execute(
+                if (previous_discharge_time < 0)
+                        log_error_errno(errno, "Error fetching battery discharge rate: %m");
+                else
+                        suspend_interval = ((previous_discharge_time * 60) - 30) * USEC_PER_MINUTE;
 
-        r = fd_wait_for_event(tfd, POLLIN, 0);
-        if (r < 0)
-                return log_error_errno(r, "Error polling timerfd: %m");
-        if (!FLAGS_SET(r, POLLIN)) /* We woke up before the alarm time, we are done. */
-                return 0;
+                log_debug("Set timerfd wake alarm to estimate battery discharge rate for %s",
+                        FORMAT_TIMESPAN(suspend_interval, USEC_PER_SEC));
+
+                timespec_store(&ts.it_value, suspend_interval);
+
+                r = timerfd_settime(tfd, 0, &ts, NULL);
+                if (r < 0)
+                        return log_error_errno(errno, "Error setting battery estimate timer: %m");
+
+                r = execute(sleep_config, SLEEP_SUSPEND, NULL);
+                if (r < 0)
+                        return r;
+
+                r = fd_wait_for_event(tfd, POLLIN, 0);
+                if (r < 0)
+                        return log_error_errno(r, "Error polling timerfd: %m");
+                if (!FLAGS_SET(r, POLLIN))  { /* We woke up before the alarm time, calculate battery discharged and estimate rate. */
+                        after_timestamp = now(CLOCK_MONOTONIC);
+
+                        current_capacity = read_battery_capacity();
+                        if (current_capacity < 0)
+                                return log_error_errno(errno, "Error fetching battery capacity: %m");
+
+                        log_debug("%d%% battery discharged in %s time",
+                                        last_capacity - current_capacity,
+                                        FORMAT_TIMESPAN(after_timestamp - before_timestamp, USEC_PER_SEC));
+                        estimated_discharge_rate = (last_capacity - current_capacity) / (before_timestamp - after_timestamp);
+                        log_debug("Battery will discharge in %s time", FORMAT_TIMESPAN(estimated_discharge_rate, USEC_PER_HOUR));
+                        r = put_battery_discharge_rate(FORMAT_TIMESPAN(estimated_discharge_rate, USEC_PER_HOUR));
+                        if (r < 0)
+                                return log_warning_errno(r, "Failed to update battery discharge rate: %m");
+
+                        return 0;
+                }
+                tfd = safe_close(tfd);
+
+                /* If woken up after alarm time, calculate battery capacity and battery capacity_level and estimate discharge rate */
+
+                current_capacity = read_battery_capacity();
+                if (current_capacity < 0)
+                        return log_error_errno(errno, "Error fetching battery capacity: %m");
+
+                estimated_discharge_rate = (last_capacity - current_capacity) / suspend_interval * USEC_PER_HOUR;
```

Comment:
some sanity checks should be done here, to ensure `suspend_interval` is a positive value, otherwise it risks dividing by zero or negative number

Reply:
Yes. added

Reply:
^~~~~~~~~~~~~~~~~~~~~~~~
../src/sleep/sleep.c:344:17: note: remove the 'if' if its condition is always true
                if(suspend_interval > 0)
                ^~~~~~~~~~~~~~~~~~~~~~~~
../src/sleep/sleep.c:274:45: note: initialize the variable 'estimated_discharge_rate' to silence this warning
                int estimated_discharge_rate;
                                            ^
                                             = 0

Probably we do not need to check suspend_interval to be greater than zero because the value is set to hibernate_delay_sec. Also build is failing because this is always true.